test/integration/cx_limit_overload_integration_test.cc

```diff
@@ -0,0 +1,213 @@
+#include <unordered_map>
+
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/overload/v3/overload.pb.h"
+
+#include "test/integration/integration.h"
+#include "test/test_common/test_runtime.h"
+
+namespace Envoy {
+
+namespace {
+
+class GlobalDownstreamCxLimitIntegrationTest
+    : public testing::TestWithParam<Network::Address::IpVersion>,
+      public BaseIntegrationTest {
+protected:
+  GlobalDownstreamCxLimitIntegrationTest()
+      : BaseIntegrationTest(GetParam(), ConfigHelper::tcpProxyConfig()) {}
+
+  void initializeOverloadManager(uint32_t max_cx) {
+    overload_manager_config_ = overloadManagerProtoConfig(max_cx);
+    config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+      *bootstrap.mutable_overload_manager() = this->overload_manager_config_;
+    });
+    initialize();
+  }
+
+  std::string counterPrefix() {
+    const bool isV4 = (version_ == Network::Address::IpVersion::v4);
+    return (isV4 ? "listener.127.0.0.1_0." : "listener.[__1]_0.");
+  }
+
+  AssertionResult waitForConnections(uint32_t expected_connections) {
+    absl::Notification num_downstream_conns_reached;
+    test_server_->server().dispatcher().post([this, &num_downstream_conns_reached,
+                                              expected_connections]() {
+      auto& overload_state = test_server_->server().overloadManager().getThreadLocalOverloadState();
+      for (int i = 0; i < 10; ++i) {
+        if (overload_state.currentResourceUsage(
+                Server::OverloadProactiveResourceName::GlobalDownstreamMaxConnections) ==
+            expected_connections) {
+          num_downstream_conns_reached.Notify();
```

Comment:
calling `Notify()` more that once on same notification is undefined behavior -- we should break out of here it we satisfied the condition